File: MandatoryAssignments/Java/Assignment1/Teque.java
import java.lang.Math;

public class Teque {

    class Node {

        int data;
        Node next;
        Node prev;

        public Node(int x) {
            data = x;
            next = null;
            prev = null;
        }

        public Node(int x, Node n, Node p) {
            data = x;
            next = n;
            prev = p;
        }
    }

    Node head;
    Node tail;
    Node middle;
    int size;

    public Teque() {
        head = null;
        tail = null;
        middle = null;
        size = 0;
    }

    public void pushBack(int x) {
        if (head == null) {
            head = new Node(x);
            tail = head;
            return;
        }

        Node current = new Node(x, null, tail);
        tail = current;
        size++;
    }

    public void pushFront(int x) {
        if (head == null) {
            head = new Node(x);
            tail = head;
            return;
        }

        Node current = new Node(x, head, null);
        head = current;
        size++;
    }


    public void pushMiddle(int x) {

        if (head == null) {
            head = new Node(x);
            tail = head;
            return;
        }

        Node current = head;
        for (int i = 0; i < Math.floor((size + 1) / 2) - 1; i++) {
            current = current.next;
        }

        Node tmp = new Node(x, current.next, current);
        Node temp = current.next;
        current.next = tmp;
        temp.prev = tmp;

        size++;
    }

    public void get(int index) {
        if (head == null) {
            System.out.println("Queue is empty");
            return;
        }

        if (index <= Math.floor((size + 1)/2)) {
            Node current = head;
            for (int i = 0; i < index; i++) {
                current = current.next;
            }

            System.out.println(current.data);
            return;
        }

        if (index > Math.floor((size + 1)/2)) {
            Node current = tail;
            for (int i = size; i > index; i--) {
                current = current.next;
            }

            System.out.println(current.data);
            return;
        }
    }
    

    public void printQueue() {
        if (head == null) {
            System.out.println("Queue is empty");
            return;
        }

        System.out.print("[ ");
        Node current = head;
        while (current != null) {
            System.out.print(current.data + " ");
            current = current.next;
        }

        System.out.println("]");
    }
}

File: MandatoryAssignments/Python/Assignment1/Teque.py
import math

class Node:

    def __init__(self, x, n, p):
        self.data = x
        self.next = n
        self.prev = p

class Teque:

    def __init__(self):
        self.head = None
        self.tail = None
        self.size = 0

    def pushBack(self, x):
        
        if self.head is None:
            self.head = Node(x, None, None)
            self.tail = self.head
            self.size += 1
            return

        current = Node(x, None, self.tail)
        self.tail = current
        self.size += 1


    def pushFront(self, x):
        
        if self.head is None:
            self.head = Node(x, None, None)
            self.tail = self.head
            self.size += 1
            return

        current = Node(x, self.tail, None)
        self.head = current
        self.size += 1

    def pushMiddle(self, x):
        if self.head is None:
            self.head = Node(x, None, None)
            self.tail = self.head
            self.size += 1
            return

        current = self.head
        for i in range(math.floor((self.size + 1)/2) - 1):
            current = current.next
            #print(current.data)
        
        tmp = Node(x, current.next, current.prev)
        temp = current.next
        current.next = tmp
        temp.prev = tmp

        self.size += 1
        

    def get(self, index):
        if self.head is None:
            print("Queue is empty")

        if index <= math.floor((self.size + 1)/2):
            current  = self.head
            for i in range(index):
                current = current.next
            
            return current.data

        if index > math.floor((self.size + 1)/2):
            current  = self.tail
            for i in range(index, 0, -1):
                current = current.prev
            
            return current.data

    def printTeque(self):
        if self.head is None:
            print("Queue is empty")
            return

        print("[ ", end="")
        current = self.head
        while (current is not None):
            print(current.data, end=" ")
            current = current.next

        print("]")

if __name__ == "__main__":

    myTeque = Teque()
    myTeque.pushBack(9)
    myTeque.pushFront(3)
    myTeque.pushMiddle(5)

    print(myTeque.get(0))
    print(myTeque.get(1))
    print(myTeque.get(2))

    myTeque.pushMiddle(1)

    myTeque.get(1)
    myTeque.get(2)
    #print(myTeque.size)

    myTeque.printTeque()